url5583f3dbb2_XOR19_XScript_tgz-pJ8-xscript_interactive_MainJ8/src/xscript/XExec.java

```diff
@@ -0,0 +1,1927 @@
+							} else {
+								call(m, right, new XArrayList<XValue>(left), null);
+							}
+						} else {
+							call(m, left, new XArrayList<XValue>(right), null);
+						}
+					} else {
+						call(m, right, new XArrayList<XValue>(left), null);
+					}
+				}
+			} else {
+				call(m, left, new XArrayList<XValue>(right), null);
+			}
+		}
+	}
+
+	private String getTypeName(XValue value) {
+		XValue type = value.getType(rt);
+		return XUtils.getDataAs(rt, type, XTypeData.class).getName();
+	}
+
+	private String __getErrorMessage2(String op, XValue left, XValue right) {
+		return "Operator " + op + " not applicable for " + getTypeName(left) + " and " + getTypeName(right);
+	}
+
+	private String __getErrorMessage1(String op, XValue value) {
+		return "Operator " + op + " not applicable for " + getTypeName(value);
+	}
+
+	private void methodNotFound(String op, XValue left, XValue right) {
+		throw new XRuntimeException("TypeError", __getErrorMessage2(op, left, right));
+	}
+
+	private void methodNotFound(String op, XValue value) {
+		throw new XRuntimeException("TypeError", __getErrorMessage1(op, value));
+	}
+
+	private void callHelper(String method, XValue... args) {
+		XValue module = rt.getBuiltinModule();
+		XValue m = XUtils.lookup(rt, module, method, XValue.REF_INTERN);
+		XObjectDataFunc func = XUtils.getDataAs(rt, m, XObjectDataFunc.class);
+		String[] paramNames = func.getParamNames();
+		if (paramNames.length != args.length)
+			throw new IllegalArgumentException();
+		call(m, XValueNull.NULL, args, false);
+	}
+
+	private void callHelper2(String method, List<XValue> params, Map<String, XValue> map) {
+		XValue module = rt.getBuiltinModule();
+		XValue m = XUtils.lookup(rt, module, method, XValue.REF_INTERN);
+		call(m, XValueNull.NULL, params, map, false);
+	}
+
+	@SuppressWarnings("unchecked")
+	private void call(XValue m, XValue thiz, XValue[] args, String[] kws, XValue list, XValue map) {
+		List<XValue> params = new ArrayList<XValue>();
+		Map<String, XValue> ma = XUtils.getDataAs(rt, map, Map.class);
+		if (ma == null) {
+			ma = new HashMap<String, XValue>();
+		} else {
+			ma = new HashMap<String, XValue>(ma);
+		}
+		if (kws == null) {
+			for (int i = 0; i < args.length; i++) {
+				params.add(args[i]);
+			}
+		} else {
+			for (int i = 0; i < args.length; i++) {
+				if (kws[i] == null) {
+					params.add(args[i]);
+				} else {
+					if (ma.containsKey(kws[i])) {
+						throw new XRuntimeException("TypeError", "Duplicated keyword '%s'", kws[i]);
+					}
+					ma.put(kws[i], args[i]);
+				}
+			}
+		}
+		if (list != null && list != XValueNull.NULL) {
+			params.addAll(XUtils.getDataAs(rt, list, Collection.class));
+		}
+		call(m, thiz, params, ma);
+	}
+
+	public void call(XValue m, XValue thiz, List<XValue> params, Map<String, XValue> map) {
+		call(m, thiz, params, map, true);
+	}
+
+	public void call(XValue m, XValue thiz, List<XValue> params, Map<String, XValue> map, boolean visibleToStacktrace) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 49. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.